File: src/api/search.service.ts
import apiClient from './axios.config';

export interface Especialidad {
    id: number;
    nombre: string;
}

export interface MedicoSearchResponse {
    id: number;
    nombres: string;
    apellidos: string;
    email: string;
    edad?: number;
    telefono?: string;
    rutaFoto?: string;
    bio?: string;
    precioConsulta?: number;
    especialidades?: Especialidad[];
    numeroColegiado?: string;
    estadoCuenta?: string;
    createdAt?: string;
}

export interface PageResponse<T> {
    content: T[];
    totalElements: number;
    totalPages: number;
    size: number;
    number: number;
}

export const searchService = {
    // Obtener todos los médicos con paginación
    getAllMedicos: async (page: number = 0, size: number = 10): Promise<PageResponse<MedicoSearchResponse>> => {
        const response = await apiClient.get(`/search/medicos?page=${page}&size=${size}`);
        return response.data;
    },

    // Buscar médicos por nombre o apellido
    searchMedicosByNombre: async (query: string, page: number = 0, size: number = 10): Promise<PageResponse<MedicoSearchResponse>> => {
        const response = await apiClient.get(`/search/medicos/search?q=${encodeURIComponent(query)}&page=${page}&size=${size}`);
        return response.data;
    },

    // Obtener detalle de un médico por ID
    getMedicoById: async (medicoId: number): Promise<MedicoSearchResponse> => {
        const response = await apiClient.get(`/search/medicos/${medicoId}`);
        return response.data;
    },

    // Filtrar médicos por especialidad
    getMedicosByEspecialidad: async (especialidadId: number, page: number = 0, size: number = 10): Promise<PageResponse<MedicoSearchResponse>> => {
        const response = await apiClient.get(`/search/medicos/especialidad/${especialidadId}?page=${page}&size=${size}`);
        return response.data;
    },

    // Filtrar médicos por rango de precios
    getMedicosByPrecioRange: async (min: number, max: number, page: number = 0, size: number = 10): Promise<PageResponse<MedicoSearchResponse>> => {
        const response = await apiClient.get(`/search/medicos/precio?min=${min}&max=${max}&page=${page}&size=${size}`);
        return response.data;
    },
};

File: backend-reference/proyecto-backend-medigo-main/src/main/java/com/example/medigo/service/SearchService.java
package com.example.medigo.service;

import com.example.medigo.domain.Medico;
import com.example.medigo.dto.response.MedicoSearchResponseDto;
import com.example.medigo.exceptions.ResourceNotFoundException;
import com.example.medigo.repository.MedicoRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.modelmapper.ModelMapper;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;

@Service
@RequiredArgsConstructor
@Slf4j
public class SearchService {

    private final MedicoRepository medicoRepository;
    private final ModelMapper modelMapper;

    @Transactional(readOnly = true)
    public Page<MedicoSearchResponseDto> getAllMedicos(Pageable pageable) {
        log.info("Buscando todos los médicos con paginación: page={}, size={}",
                pageable.getPageNumber(), pageable.getPageSize());

        Page<Medico> medicos = medicoRepository.findAll(pageable);
        return medicos.map(medico -> {
            MedicoSearchResponseDto dto = modelMapper.map(medico, MedicoSearchResponseDto.class);
            dto.setNumeroColegiado(medico.getNumeroColegiado());
            return dto;
        });
    }

    @Transactional(readOnly = true)
    public Page<MedicoSearchResponseDto> searchMedicosByNombre(String query, Pageable pageable) {
        log.info("Buscando médicos por nombre/apellido: query='{}', page={}, size={}",
                query, pageable.getPageNumber(), pageable.getPageSize());

        Page<Medico> medicos = medicoRepository.findByNombresContainingIgnoreCaseOrApellidosContainingIgnoreCase(
                query, query, pageable);
        return medicos.map(medico -> {
            MedicoSearchResponseDto dto = modelMapper.map(medico, MedicoSearchResponseDto.class);
            dto.setNumeroColegiado(medico.getNumeroColegiado());
            return dto;
        });
    }

    @Transactional(readOnly = true)
    public MedicoSearchResponseDto getMedicoByEmail(String email) {
        log.info("Buscando médico por email: {}", email);

        Medico medico = medicoRepository.findByEmail(email)
                .orElseThrow(() -> new ResourceNotFoundException("Médico no encontrado con email: " + email));

        MedicoSearchResponseDto dto = modelMapper.map(medico, MedicoSearchResponseDto.class);
        dto.setNumeroColegiado(medico.getNumeroColegiado());
        return dto;
    }

    @Transactional(readOnly = true)
    public MedicoSearchResponseDto getMedicoById(Long medicoId) {
        log.info("Buscando médico por ID: {}", medicoId);

        Medico medico = medicoRepository.findById(medicoId)
                .orElseThrow(() -> new ResourceNotFoundException("Médico no encontrado con ID: " + medicoId));

        log.info("Médico encontrado: {} {}, numeroColegiado en entidad: {}",
                medico.getNombres(), medico.getApellidos(), medico.getNumeroColegiado());

        MedicoSearchResponseDto dto = modelMapper.map(medico, MedicoSearchResponseDto.class);
        dto.setNumeroColegiado(medico.getNumeroColegiado());

        log.info("DTO creado, numeroColegiado en DTO: {}", dto.getNumeroColegiado());

        return dto;
    }

    @Transactional(readOnly = true)
    public Page<MedicoSearchResponseDto> getMedicosByEspecialidad(Long especialidadId, Pageable pageable) {
        log.info("Buscando médicos por especialidad ID: {}, page={}, size={}",
                especialidadId, pageable.getPageNumber(), pageable.getPageSize());

        Page<Medico> medicos = medicoRepository.findByEspecialidadesId(especialidadId, pageable);
        return medicos.map(medico -> {
            MedicoSearchResponseDto dto = modelMapper.map(medico, MedicoSearchResponseDto.class);
            dto.setNumeroColegiado(medico.getNumeroColegiado());
            return dto;
        });
    }

    @Transactional(readOnly = true)
    public Page<MedicoSearchResponseDto> getMedicosByEspecialidadNombre(String nombreEspecialidad, Pageable pageable) {
        log.info("Buscando médicos por especialidad nombre: {}, page={}, size={}",
                nombreEspecialidad, pageable.getPageNumber(), pageable.getPageSize());

        Page<Medico> medicos = medicoRepository.findByEspecialidadNombre(nombreEspecialidad, pageable);
        return medicos.map(medico -> {
            MedicoSearchResponseDto dto = modelMapper.map(medico, MedicoSearchResponseDto.class);
            dto.setNumeroColegiado(medico.getNumeroColegiado());
            return dto;
        });
    }

    @Transactional(readOnly = true)
    public Page<MedicoSearchResponseDto> getMedicosByPrecioRange(BigDecimal minPrecio,
            BigDecimal maxPrecio,
            Pageable pageable) {
        log.info("Buscando médicos por rango de precio: min={}, max={}, page={}, size={}",
                minPrecio, maxPrecio, pageable.getPageNumber(), pageable.getPageSize());

        Page<Medico> medicos = medicoRepository.findByPrecioConsultaBetween(minPrecio, maxPrecio, pageable);
        return medicos.map(medico -> {
            MedicoSearchResponseDto dto = modelMapper.map(medico, MedicoSearchResponseDto.class);
            dto.setNumeroColegiado(medico.getNumeroColegiado());
            return dto;
        });
    }
}
